crypto.go

```diff
@@ -209,17 +238,47 @@ type sharedSecretGenerator interface {
 	generateSharedSecret(dhKey *btcec.PublicKey) (Hash256, error)
 }
 
-// generateSharedSecret generates the shared secret by given ephemeral key.
-func (r *Router) generateSharedSecret(dhKey *btcec.PublicKey) (Hash256, error) {
+// generateSharedSecret generates the shared secret by given
+// ephemeral key. If a blindingPoint is provided then it is used to tweak our
+// private key before creating the shared secret with the ephemeral key.
+func (r *Router) generateSharedSecret(dhKey *btcec.PublicKey,
+	blindingPoint *btcec.PublicKey) (Hash256, error) {
+
+	// If no blinding point is provided, then the un-tweaked dhKey can
+	// be used to derive the shared secret
+	if blindingPoint == nil {
+		return sharedSecret(r.onionKey, dhKey)
+	}
+
+	// We use the blinding point to calculate the blinding factor that the
+	// receiver used with us so that we can use it to tweak our priv key.
+	// The sender would have created their shared secret with our blinded
+	// pub key.
+	ssReceiver, err := sharedSecret(r.onionKey, blindingPoint)
+	if err != nil {
+		return Hash256{}, err
+	}
+
+	blindingFactorBytes := generateKey(routeBlindingHMACKey, &ssReceiver)
+	var blindingFactor btcec.ModNScalar
+	blindingFactor.SetBytes(&blindingFactorBytes)
+
+	ephemeral := blindGroupElement(dhKey, blindingFactor)
```

Comment:
yes indeed :) i have added a comment to point this out

Reply:
Apologies for any confusion, Ziggie. It's been a while since I've looked at this, but I'm pretty sure that when I say `blinding_factor` here I am referring to Elle's variable `blindingFactor := HMAC256("blinded_node_id", ss(i))`. Which you are correct does differ from how the term is used in the spec document. Brought that up with Elle briefly [here](https://github.com/lightningnetwork/lightning-onion/pull/57#discussion_r932640095)

Also, looks like my link may have gotten rug pulled by a change to t-bast's underlying spec PR. I usually pick an older, non-master release or tag so that links point to the expected line, but I don't know how to do that for an actively changing PR :/  In any case, the section this used to link to is now [here](https://github.com/lightning/bolts/pull/765/files#diff-ad65f0beaac5cef88f5fd7a8b9ca36cbc5a790f4815b4b50b9ea794a55aaf012R510-R513).